Reconstruct scopes

taivu1998 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42133 Get, save, and load module information for each operator
* **#41615 Reconstruct scopes**

Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

Commit:
Reconstruct scopes

[ghstack-poisoned]

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]

kimishpatel commented on torch/csrc/jit/passes/reconstruct_scopes.cpp:
```diff
@@ -0,0 +1,105 @@
+  void run();
+
+ private:
+  const Module* root_module;
+  Graph* graph;
```
Is there a reason why we are storing Module and Graph as pointers? Can we instead store const refs? Like `const Module& root_module`?

kimishpatel replied:
I guess we update graph nodes. Is that why it is not const?

kimishpatel replied:
Actually I missed something. We dont want to store pointers/references to something that may have shorter lifetime than the object itself. If we construct ReconstructScopesPass and pass it around while the module/graph that was passed in is destroyed, we are holding onto something that is not valid.
Either we ensure that they remain valid through the lifetime of ReconstructScopesPass, or we should just copy.

kimishpatel replied:
OK, I understand. We do need original module as we are mapping Function pointers.

kimishpatel commented on torch/csrc/jit/passes/reconstruct_scopes.cpp:
```diff
@@ -0,0 +1,105 @@
+};
+
+void ReconstructScopesPass::constructFunctionToModuleMap(const Module& module) {
+  for (auto& method : module.get_methods()) {
```
`const auto&`

kimishpatel commented on torch/csrc/jit/passes/reconstruct_scopes.cpp:
```diff
@@ -0,0 +1,105 @@
+  for (auto& method : module.get_methods()) {
+    func_to_module[&method.function()] = module._ivalue();
+  }
+  for (Module m : module.children()) {
```
`const auto&`

kimishpatel commented on torch/csrc/jit/passes/reconstruct_scopes.cpp:
```diff
@@ -0,0 +1,105 @@
+    const Module& module,
+    const std::string& prefix) {
+  module_names[module._ivalue()] = prefix;
+  for (NameModule s : module.named_children()) {
```
`const ref`

kimishpatel commented on torch/csrc/jit/passes/reconstruct_scopes.cpp:
```diff
@@ -0,0 +1,105 @@
+  }
+}
+
+std::string ReconstructScopesPass::getScopeString(Function* f) const {
```
`const Function*`?

kimishpatel commented on torch/csrc/jit/passes/reconstruct_scopes.cpp:
```diff
@@ -0,0 +1,105 @@
+  return module_names.at(m) + "." + f->name();
+}
+
+void ReconstructScopesPass::visitNode(Node* n) {
```
const Node

taivu1998 replied:
I did try `const` with `Node*`, `Block*`, and `Function*`, but it raised compiler errors because we modified the objects such as `n->setScope(...)`

kimishpatel replied:
Aah. Ok nevermind.

kimishpatel commented on torch/csrc/jit/passes/reconstruct_scopes.cpp:
```diff
@@ -0,0 +1,105 @@
+    return;
+  }
+  ScopePtr sc = c10::make_intrusive<Scope>();
+  for (auto frame : (*n->callstack())->vec()) {
```
`const auto& frame`

kimishpatel commented on torch/csrc/jit/passes/reconstruct_scopes.cpp:
```diff
@@ -0,0 +1,105 @@
+  GRAPH_UPDATE("Updated node: ", *n);
+}
+
+void ReconstructScopesPass::visitBlock(Block* b) {
```
`const Block`

kimishpatel reviewed (changes requested):
Left a few comments. Please acknowledge that this was basically lifted from an earlier implementation by @ZolotukhinM.

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/reconstruct_scopes.cpp:
```diff
@@ -0,0 +1,105 @@
+  Graph& graph;
+  std::string prefix;
+
+  std::unordered_map<Function*, ModulePtr> func_to_module;
```
Nit: names of private fields usually end with an underscore.

ZolotukhinM commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -3,6 +3,7 @@
 #include <c10/util/Exception.h>
 #include <torch/csrc/jit/ir/type_hashing.h>
 #include <torch/csrc/jit/passes/inliner.h>
+#include <torch/csrc/jit/passes/reconstruct_scopes.h>
```
Nit: this doesn't seem to be necessary.

ZolotukhinM reviewed (approved):
Looks good!

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]

Commit:
Update on "Reconstruct scopes"


Differential Revision: [D22611331](https://our.internmc.facebook.com/intern/diff/D22611331)

[ghstack-poisoned]
